### PR TITLE
Custom CheckStyle Rule: HttpPipelinePolicy 

### PR DESCRIPTION
**Pull request by mssfang**
Any class that implements the HttpPipelinePolicy interface should:

-  Must be a public class.
-  Not in an implementation package or sub-package.

### CODE REVIEW

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Http Pipeline Policy Checks
+ * Any class that implements the HttpPipelinePolicy interface should:
+ *  <ol>
+ *    <li>must be a public class</li>
```
Comment (conniey) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java]:
```suggestion
 *    <li>Be a public class.</li>
```

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ *    <li>not in an implementation package or sub-package</li>
```
Comment (conniey) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java]:
```suggestion
 *    <li>Not live in the implementation package or any of its sub-packages.</li>
```

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ *  </ol>
+ */
+public class HttpPipelinePolicyCheck extends AbstractCheck {
+    private static final String HTTP_PIPELINE_POLICY = "HttpPipelinePolicy";
+    private boolean isImplePackage;
```
Comment (conniey) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java]: isImplementationPackage. This is an unnecessary abbreviation.

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    @Override
+    public void visitToken(DetailAST ast) {
+
+        switch (ast.getType()) {
+            case TokenTypes.PACKAGE_DEF:
+                String packageName = FullIdent.createFullIdentBelow(ast).getText();
+                isImplePackage = packageName.contains("implementation");
+                break;
+            case TokenTypes.CLASS_DEF:
+                String className = ast.findFirstToken(TokenTypes.IDENT).getText();
```
Comment (conniey) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java]: You can save yourself some cost by moving this line to after line 60. You don't need the className if this class doesn't implement anything.

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                // Get all interfaces name
+                DetailAST impleClauseToken = ast.findFirstToken(TokenTypes.IMPLEMENTS_CLAUSE);
+                // Skip check if the class doesn't implement any interface
+                if (impleClauseToken == null) {
+                    return;
+                }
+                Set<String> interfaces = new HashSet<>();
```
Comment (conniey) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java]: Why create a set for this? You can simplify this by using 
```java
Optional<DetailAST> policy = TokenUtils.findFirstTokenByPredicate(impleClauseToken, node -> {
  return node.getType() == TokenTypes.IDENT && HTTP_PIPELINE_POLICY.equals(node.getText());
});
```

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                    if (node.getType() == TokenTypes.IDENT) {
+                        interfaces.add(node.getText());
+                    }
+                }
+                // Skip check if the class doesn't implement HTTP_PIPELINE_POLICY
+                if (!interfaces.contains(HTTP_PIPELINE_POLICY)) {
+                    return;
+                }
+
+                DetailAST modifiersToken = ast.findFirstToken(TokenTypes.MODIFIERS);
```
Comment (conniey) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java]: Replace this with: `CheckUtil.getAccessModifierFromModifiersToken`

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            case TokenTypes.CLASS_DEF:
+                String className = ast.findFirstToken(TokenTypes.IDENT).getText();
+                // Get all interfaces name
+                DetailAST impleClauseToken = ast.findFirstToken(TokenTypes.IMPLEMENTS_CLAUSE);
+                // Skip check if the class doesn't implement any interface
+                if (impleClauseToken == null) {
+                    return;
+                }
+                Set<String> interfaces = new HashSet<>();
+                for (DetailAST node = impleClauseToken.getFirstChild(); node != null; node = node.getNextSibling()) {
```
Comment (vhvb1989) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java]: You can update this to use the TokenUtil for simplicity
```
TokenUtil.forEachChild(impleClauseToken, TokenTypes.IDENT, node -> interfaces.add(node .getText());
```

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Override
+    public void visitToken(DetailAST ast) {
+
+        switch (ast.getType()) {
+            case TokenTypes.PACKAGE_DEF:
+                String packageName = FullIdent.createFullIdentBelow(ast).getText();
+                isImplePackage = packageName.contains("implementation");
+                break;
+            case TokenTypes.CLASS_DEF:
```
Comment (vhvb1989) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java]: Can you move the content of this case to a new method?  
I think it is fine to have 2-4 lines max or cases, else, better to have a method call to describe that will be done

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                }
+                // Skip check if the class doesn't implement HTTP_PIPELINE_POLICY
+                if (!interfaces.contains(HTTP_PIPELINE_POLICY)) {
+                    return;
+                }
+
+                DetailAST modifiersToken = ast.findFirstToken(TokenTypes.MODIFIERS);
+                // Public class check
+                if (!modifiersToken.branchContains(TokenTypes.LITERAL_PUBLIC)) {
+                    log(modifiersToken, String.format("Class ''%s'' implements ''%s'' should be a public class", className, HTTP_PIPELINE_POLICY));
```
Comment (vhvb1989) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java]: add `and`  or a dot `.` before `should be a pub...` to better explain

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+                }
+                // Implementation and sub-package check
+                if (isImplePackage) {
+                    log(ast, String.format("Class ''%s'' implements ''%s'' should not be a implementation package or sub-package of it", className, HTTP_PIPELINE_POLICY));
```
Comment (vhvb1989) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java]: same here..  or change `implements` to `implementing`

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java (hunk 7 above)
Comment (mssfang) [reply eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/HttpPipelinePolicyCheck.java]: outdated

File: eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml [hunk 11]
```diff
@@ -96,12 +96,14 @@
   <suppress checks="com.azure.tools.checkstyle.checks.ExternalDependencyExposedCheck" files=".*[/\\]com[/\\]microsoft[/\\].*"/>
   <suppress checks="com.azure.tools.checkstyle.checks.ServiceClientInstantiationCheck" files=".*[/\\]com[/\\]microsoft[/\\].*"/>
   <suppress checks="com.azure.tools.checkstyle.checks.ThrownClientLoggerCheck" files=".*[/\\]com[/\\]microsoft[/\\].*"/>
+  <suppress checks="com.azure.tools.checkstyle.checks.HttpPipelinePolicyCheck" files=".*[/\\]com[/\\]microsoft[/\\].*"/>
```
Comment (conniey) [inline eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml]: We have a lot of the lines for suppressions of files that are for our Track 1 APIs (com.microsoft.*).  I think a lot of these lines could be removed and replaced with a single line. (I read that "checks" can take a regex,)
```xml
<suppress checks="com\.azure\.tools\.checkstyle\.checks\.*" files=".*[/\\]com[/\\]microsoft[/\\].*"/>
```
Thoughts,  @JonathanGiles? Or do you prefer it be explicit for all these custom checkstyle rules.

Supression Files Docs: https://checkstyle.sourceforge.io/config_filters.html#SuppressionFilter_Notes

Comment (JonathanGiles) [reply eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml]: I would love to see this change. Please make it so @mssfang.

**Pull request merged**